library/std/src/sys_common/condvar/check.rs

```diff
@@ -24,7 +24,7 @@ impl SameMutexCheck {
     }
     pub fn verify(&self, mutex: &MovableMutex) {
         let addr = mutex.raw() as *const imp::Mutex as *const () as *mut _;
-        match self.addr.compare_exchange(ptr::null_mut(), addr, Ordering::SeqCst, Ordering::SeqCst)
+        match self.addr.compare_exchange(ptr::null_mut(), addr, Ordering::Relaxed, Ordering::Relaxed)
```

Comment:
So, this is fine, but looks wrong.

That is, if we ever would read through `self.addr` to get the underlying mutex, then Relaxed would be insufficient for this, due to it not ensuring data-dependencies (such as the writes to the value behind the pointer) are correctly ordered with respect to the load/store of the pointer. E.g. you'd need acquire/release semantics, (although in principal if we had consume, I suppose consume/release would be fine). 

This would only matter on something like the Alpha though (and for better or worse, that's one of the targets that I've heard people want to use with rustc-codegen-gcc eventually).

Anyway, we don't do that here, so Relaxed should be fine. But it looks wrong (Relaxed is *almost* always wrong with AtomicPtr), so I think it needs a comment. Perhaps something along the lines of "Relaxed is okay here because we never read through `self.addr`, and only use it to compare addresses", but you can feel free to reword if you'd like

Reply:
Great point. I’ll update to add a comment.